### PR TITLE
bug 1166448 - support golden ami creation for win 2008 bld/try

### PR DESCRIPTION
**Pull request by grenade**
These changes introduce 2 machine and userdata configs for windows 2008 build and try golden images. cloudtools/aws/ami.py was also modified to remove unix path cleanup which is not applicable on windows.

### CODE REVIEW

File: cloudtools/aws/ami.py [hunk 1]
```diff
@@ -12,28 +12,29 @@
...
-            "usr/sbin/fake_puppet.sh", mirror_local_mode=True)
-        # replace puppet init with our script
-        if distro == "ubuntu":
-            put("%s/fake_puppet.conf" % AMI_CONFIGS_DIR,
-                "etc/init/puppet.conf", mirror_local_mode=True)
-            run("echo localhost > etc/hostname")
-        else:
-            run("ln -sf /usr/sbin/fake_puppet.sh etc/init.d/puppet")
-            run('echo "NETWORKING=yes" > etc/sysconfig/network')
+    if distro != "win2008":
```
Comment (rail) [inline cloudtools/aws/ami.py]: Hmm, I wonder if we should move this condition to the configs or use something like if distro.startswith("win")...

File: configs/bld-win64 [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        "ami": "ami-fe051b96",
+        "subnet_ids": [ "subnet-2ba98340", "subnet-2da98346", "subnet-22a98349", "subnet-0822004e", "subnet-2da98346", "subnet-5bc7c62f", "subnet-7091d358" ],
+        "security_group_ids": [
+            "sg-18a07677",
+            "sg-f3927c9c"
+        ],
+        "instance_type": "m1.medium",
+        "distro": "win2008",
+        "ssh_key": "aws-releng",
+        "user_data_file": "/builds/aws_manager/cloud-tools/configs/bld-win64.cmd.user_data",
```
Comment (rail) [inline configs/bld-win64]: The path shouldn't be absolute. It won't work on a local machine or if we decide to change the base directory.

File: configs/bld-win64 [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        "ami": "ami-fe051b96",
+        "subnet_ids": ["subnet-d748dabe", "subnet-a848dac1", "subnet-ad48dac4", "subnet-c74f48b3"],
+        "security_group_ids": [
+            "sg-d5617cb9",
+            "sg-84beade6"
+        ],
+        "instance_type": "m1.medium",
+        "distro": "win2008",
+        "ssh_key": "aws-releng",
+        "user_data_file": "/builds/aws_manager/cloud-tools/configs/bld-win64.cmd.user_data",
```
Comment (rail) [inline configs/bld-win64]: the same here

File: configs/try-win64 [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        "ami": "ami-fe051b96",
+        "subnet_ids": [ "subnet-27a9834c", "subnet-39a98352", "subnet-3ea98355", "subnet-93b285e7", "subnet-e5bacacd", "subnet-cd83d28b" ],
+        "security_group_ids": [
+            "sg-18a07677",
+            "sg-f3927c9c"
+        ],
+        "instance_type": "m1.medium",
+        "distro": "win2008",
+        "ssh_key": "aws-releng",
+        "user_data_file": "/builds/aws_manager/cloud-tools/configs/try-win64.cmd.user_data",
```
Comment (rail) [inline configs/try-win64]: and here

File: configs/try-win64 [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        "ami": "ami-fe051b96",
+        "subnet_ids": ["subnet-ae48dac7", "subnet-a348daca", "subnet-a448dacd", "subnet-72b68206"],
+        "security_group_ids": [
+            "sg-d5617cb9",
+            "sg-84beade6"
+        ],
+        "instance_type": "m1.medium",
+        "distro": "win2008",
+        "ssh_key": "aws-releng",
+        "user_data_file": "/builds/aws_manager/cloud-tools/configs/try-win64.cmd.user_data",
```
Comment (rail) [inline configs/try-win64]: and here

File: configs/try-win64.cmd.user_data [hunk 6]
```diff
@@ -0,0 +1,8 @@
+<script>
+sc delete puppet
+cscript.exe "C:\ProgramData\Puppetlabs\puppet\var\puppettize_TEMP.vbs"
+cd "C:\Program Files (x86)\Puppet Labs\Puppet\bin"
+cmd /c puppet agent --test --environment=mcornmesser --server=releng-puppet2.srv.releng.scl3.mozilla.com
+schtasks /delete /TN RUNPUPPET /F
+shutdown /s /t 0 /c "userdata run complete" /f /d p:4:1
+</script>
```
Comment (rail) [inline configs/try-win64.cmd.user_data]: I'm not an expert to review these files, but they look sane to me. Do you still need to use --environment here?

File: cloudtools/aws/ami.py (hunk 1 above)
Comment (djmitche) [inline cloudtools/aws/ami.py]: preferably in configs, yeah

**Comment by amyrrich** [comment]
We should also change the file names to match the machines names, e.g. b-2008 and b-2008.cmd.user_data, try-2008 and try-2008.cmd.user_data (or whatever we wind up calling try).

Hopefully mcornmesser is going to land all of his patches soon, so when we actually submit this patch, it'll just be using the production puppet environment (no need to specify)

**Comment by amyrrich** [comment]
It also looks like these still reference the configurations for the old, defunct tst-w64 stuff in ami_configs. We need to clean all that out and make new configs for b-2008 and try-2008 that have the correct ami-ids, security groups, instance types, domains, etc.

File: cloudtools/aws/ami.py [hunk 7]
```diff
@@ -12,28 +12,29 @@
...
+            "root/*.sh",
+            "root/*.log",
+            "root/userdata",
+            "var/lib/puppet",
+            "etc/init.d/puppet"
+        ]
+        with cd(mount_point):
+            for e in remove + remove_extra:
+                run('rm -rf %s' % (e,))
+            run("sed -i -e 's/127.0.0.1.*/127.0.0.1 localhost/g' etc/hosts")
```
Comment (petemoore) [inline cloudtools/aws/ami.py]: the literal dots should be escaped - e.g. 127\.0\.0\.1

File: cloudtools/aws/ami.py [hunk 8]
```diff
@@ -12,28 +12,29 @@
...
+        remove = [
+            "root/*.sh",
+            "root/*.log",
+            "root/userdata",
+            "var/lib/puppet",
+            "etc/init.d/puppet"
+        ]
+        with cd(mount_point):
+            for e in remove + remove_extra:
+                run('rm -rf %s' % (e,))
```
Comment (petemoore) [inline cloudtools/aws/ami.py]: "rm -rf '%s'" might be better here - otherwise the shell could expand special characters in the filename, and also it will treat %s as a single argument if a file name has spaces

File: configs/bld-win64.cmd.user_data [hunk 9]
```diff
@@ -0,0 +1,8 @@
+<script>
+sc delete puppet
+cscript.exe "C:\ProgramData\Puppetlabs\puppet\var\puppettize_TEMP.vbs"
+cd "C:\Program Files (x86)\Puppet Labs\Puppet\bin"
+cmd /c puppet agent --test --environment=mcornmesser --server=releng-puppet2.srv.releng.scl3.mozilla.com
+schtasks /delete /TN RUNPUPPET /F
+shutdown /s /t 0 /c "userdata run complete" /f /d p:4:1
+</script>
```
Comment (petemoore) [inline configs/bld-win64.cmd.user_data]: regular batch scripts don't exit if a command fails (there is nothing like `set -e` like in bash) - do we know if one of the earlier steps fails, whether subsequent steps will run? In other words, is this copied to target machine and then executed? If so, maybe we need checking of %errorlevel%?

File: cloudtools/aws/ami.py (hunk 8 above)
Comment (djmitche) [inline cloudtools/aws/ami.py]: If you're going to worry about quoting, the best-in-class approach is [`pipes.quote`](https://docs.python.org/2/library/pipes.html#pipes.quote) (or `shlex.quote` in python3.3)

File: cloudtools/aws/ami.py (hunk 1 above)
Comment (grenade) [inline cloudtools/aws/ami.py]: rail or djmitche: can you expand on how this could/should be handled in config? All the distro specific stuff I see, seems to be handled with `if distro == ...`. I'm not clear on what is needed to fix this.

File: cloudtools/aws/ami.py (hunk 8 above)
Comment (grenade) [inline cloudtools/aws/ami.py]: The changes suggested may be better addressed in another PR as this one only deals with https://bugzilla.mozilla.org/show_bug.cgi?id=1166448

File: cloudtools/aws/ami.py (hunk 7 above)
Comment (grenade) [inline cloudtools/aws/ami.py]: The changes suggested may be better addressed in another PR as this one only deals with https://bugzilla.mozilla.org/show_bug.cgi?id=1166448

File: cloudtools/aws/ami.py (hunk 1 above)
Comment (djmitche) [inline cloudtools/aws/ami.py]: /me leaves to @rail

Comment (rail) [inline cloudtools/aws/ami.py]: We can add another config variable to distinguish windows and non-window platforms. Something like platform_type = windows|linux|osx maybe? Or we can use the same approach as in https://github.com/mozilla/build-cloud-tools/blob/60fc80ac75726b9dcefedc86a48f90f0f2aeed1c/cloudtools/aws/instance.py#L107

Comment (grenade) [inline cloudtools/aws/ami.py]: I went with the second option (https://github.com/mozilla/build-cloud-tools/commit/e19cac64c77c668257a483871bcf2450e30a0ed0), since we'd still need if/else in the code and I'd like to minimise the change/churn in this PR.

File: configs/try-win64.cmd.user_data (hunk 6 above)
Comment (grenade) [inline configs/try-win64.cmd.user_data]: `--environment=mcornmesser` removed, as per Amy's comments below.

**Comment by grenade** [comment]
@rail I believe we're ready to merge if you are happy with the changes. I have squashed all commits (including those from @amyrrich), for this PR, into one, to tidy up the merge.

**Comment by rail** [comment]
:shipit:
